pkg/apis/kops/validation/legacy.go

```diff
@@ -291,6 +291,36 @@ func ValidateCluster(c *kops.Cluster, strict bool) field.ErrorList {
 					}
 				}
 			}
+
+			// @ check that NodeLocalDNS addon is configured correctly
+			if c.Spec.KubeDNS.NodeLocalDNS != nil && c.Spec.KubeDNS.NodeLocalDNS.Enabled {
+				if c.Spec.KubeDNS.Provider != "CoreDNS" {
+					allErrs = append(allErrs, field.Forbidden(fieldSpec.Child("kubeDNS", "provider"), "KubeDNS provider must be set to CoreDNS if NodeLocalDNS addon is enabled"))
+				}
+
+				if c.Spec.KubeDNS.NodeLocalDNS.LocalIP == "" {
+					if c.Spec.Kubelet != nil && c.Spec.Kubelet.ClusterDNS == "" {
```

Comment:
Looks like at this point `c.Spec.Kubelet.ClusterDNS` is set to `100.64.0.10` so one will never enter this block.
I think a better check is `c.Spec.Kubelet.ClusterDNS != c.Spec.KubeDNS.NodeLocalDNS.LocalIP`.

Reply:
> Looks like at this point c.Spec.Kubelet.ClusterDNS is set to 100.64.0.10 so one will never enter this block.

Could you elaborate here the why?

Also I do already that check here https://github.com/kubernetes/kops/pull/8780/files/429f0fec9701236f79d9608506b1af91b1709a28#diff-33635946dcf9aa666832a9236f8f3349R318

Reply:
I didn't look much into the details. I just noticed that I did not get the expected error when using cilium and not changing kubelet.ClusterDNS.

Reply:
But yeah it makes sense somewhere `100.64.0.10` it is set up. I've run a global lookup of it in the codebase and just found that it is generated as output. So yeah somewhere is set.

Reply:
I've updated here the logic.